ex/lib/rpc_server/router.ex

```diff
@@ -0,0 +1,92 @@
+defmodule Ama.RpcServer do
+  use Plug.Router
+  plug :fetch_auth
+  plug :match
+  plug :dispatch
+
+
+  defp fetch_auth(conn, _opts) do
+    username = Application.get_env(:ama, :rpc_user) || ""
+    password = Application.get_env(:ama, :rpc_password) || ""
+  
+    case get_req_header(conn, "authorization") do
+      ["Basic " <> _] ->
+        Plug.BasicAuth.basic_auth(conn, username: username, password: password)
+      _ ->
+        # No authentication header, assume empty username and password
+        auth_header = "Basic " <> Base.encode64(":")
+        conn
+        |> put_req_header("authorization", auth_header)
+        |> Plug.BasicAuth.basic_auth(username: username, password: password)
+    end
+  end
+  post "/" do
+    {:ok, body, _conn} = Plug.Conn.read_body(conn)
+
+    case Jason.decode(body) do
+      {:ok, %{"method" => method} = params} ->
+        result = call_dynamic_method(method, params)
```

Comment:
instead of this complex function to be clear you can do
```
%{"module" => module, "function"=> function, "args"=> args} = params
module = String.to_existing_atom(module)
function = String.to_existing_atom(function)
raw_result = try do
  module.function(args)
catch
  e,r ->
    IO.inspect {:rpc_error, e, r, __STACKTRACE__}
    %{error: e, reason: r}
end
Jason.encode!(raw_result)
```

Something like this, its much more dynamic and saves a ton of lines of code